torch/csrc/lazy/core/view_ops/narrow.cpp

```diff
@@ -16,7 +16,7 @@ Narrow::Narrow(
           MHash(base_indices, sizes)),
       base_indices_(base_indices.begin(), base_indices.end()),
       sizes_(sizes.begin(), sizes.end()) {
-  SetShapeDeferred([&]() {
+  addShapeFromFunction([&]() {
```

Comment:
I guess looking at it from the call-site perspective, this function call does actually set the node shape.  It may either compute it or use the cached version.

SetShapeDeferred was always a confusing name for people.  I think the 'SetShape' part is correct, but what is confusing is why it is deferred or why we have special behavior at all.  'FromFunction' is kind of obvious because you can see the function, but it still doesn't explain the unclear part, that backend may cache it.  I wonder if `setShapeMaybeCached` is a better name

Reply:
I don't know if `setShapeMaybeCached` is the best name here as its not conditionally cached (which is what that name implies to me). Its always cached and only computed if never previously cached. Since we're hiding that caching logic in the backend and not in the node, I don't think it makes sense to limit the scope of this function to just be caching. A backend could theoretically do anything it wants with that shape function so long as it returns some shape.

So, I think `computeShape` still works as a name. The shape is being computed, but a backend can choose to return a cached computation should they choose.

Alternatively, if we do want to push heavily that its purpose is strictly to facilitate caching, we can name it something like `setShapeCached(default_shape_fn)` where `default_shape_fn` is called if never called before.

Thoughts @wconstab?

Reply:
yea, i agree that 'maybeCached' is imprecise as who knows what the backend will do.  I am ok with computeShape

Reply:
okay, one final change. I think changing the name from `addShapeFromFunction` to `addComputedShape` makes things inline with the other `computeShape` function. If it would be better, I'd also be okay with calling this `setComputedShape` (although adding is technically a bit more accurate)